### DISCUSSION TITLE
Is the `bucket notification` planned for production?

### DISCUSSION THREAD

**Discussion by microyahoo** [rook/rook#11484] [Q&A] [answered]
https://rook.io/docs/rook/v1.10/Storage-Configuration/Object-Storage-RGW/ceph-object-bucket-notifications/
I have noticed object bucket notification feature is experimental. Is there any plan for production? Thanks.

**Comment by travisn** [comment]
The bucket notifications have been in since the v1.8 release, so we can certainly evaluate moving the feature to GA. @microyahoo Have you tested the feature, or do you have any feedback on the feature? 

@yuvalif Anything remaining from your perspective for bucket notifications in Rook moving to GA?

**Reply by microyahoo** [reply]
Yes, I have tried this feature for a while and found a issue. The following is the detailed steps.
1. refer to [CRD](https://rook.io/docs/rook/v1.10/Storage-Configuration/Object-Storage-RGW/ceph-object-bucket-notifications/#cephbuckettopic-custom-resource) to create `CephBucketTopic` 
```yaml
[root@node01 smd]# kubectl get cephbuckettopics.ceph.rook.io -n rook-ceph c-dev-prod-bucket-notification -o yaml
apiVersion: ceph.rook.io/v1
kind: CephBucketTopic
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"ceph.rook.io/v1","kind":"CephBucketTopic","metadata":{"annotations":{},"name":"c-dev-prod-bucket-notification","namespace":"rook-ceph"},"spec":{"endpoint":{"kafka":{"ackLevel":"broker","disableVerifySSL":true,"uri":"kafka://prod-kafka.deeproute.cn:9092","useSSL":false}},"objectStoreName":"os-dsglczutvqsgowpz","objectStoreNamespace":"rook-ceph","opaqueData":"liangsu@deeproute.ai","persistent":true}}
  creationTimestamp: "2022-12-12T03:09:17Z"
  finalizers:
  - cephbuckettopic.ceph.rook.io
  generation: 2
  name: c-dev-prod-bucket-notification
  namespace: rook-ceph
  resourceVersion: "26666098"
  uid: ceee143d-f02c-49d5-a201-971f5fac41d0
spec:
  endpoint:
    kafka:
      ackLevel: broker
      disableVerifySSL: true
      uri: kafka://prod-kafka.deeproute.cn:9092
  objectStoreName: os-dsglczutvqsgowpz
  objectStoreNamespace: rook-ceph
  opaqueData: xxxx
  persistent: true   # <- set persistent to true
status:
  ARN: arn:aws:sns:os-dsglczutvqsgowpz::c-dev-prod-bucket-notification
  observedGeneration: 1
  phase: Ready
```
```console
root@node09:/# radosgw-admin topic get --topic c-dev-prod-bucket-notification
{
    "topic": {
        "user": "",
        "name": "c-dev-prod-bucket-notification",
        "dest": {
            "bucket_name": "",
            "oid_prefix": "",
            "push_endpoint": "kafka://prod-kafka.deeproute.cn:9092",
            "push_endpoint_args": "Attributes.entry.1.key=OpaqueData&Attributes.entry.1.value=liangsu@deeproute.ai&Attributes.entry.2.key=kafka-ack-level&Attributes.entry.2.value=broker&Attributes.entry.3.key=persistent&Attributes.entry.3.value=true&Attributes.entry.4.key=push-endpoint&Attributes.entry.4.value=kafka://prod-kafka.deeproute.cn:9092&Attributes.entry.5.key=use-ssl&Attributes.entry.5.value=false&Attributes.entry.6.key=verify-ssl&Attributes.entry.6.value=false&OpaqueData=xxxi&Version=2010-03-31&kafka-ack-level=broker&persistent=true&push-endpoint=kafka://prod-kafka.deeproute.cn:9092&use-ssl=false&verify-ssl=false",
            "push_endpoint_topic": "c-dev-prod-bucket-notification",
            "stored_secret": "false",
            "persistent": "true"
        },
        "arn": "arn:aws:sns:os-dsglczutvqsgowpz::c-dev-prod-bucket-notification",
        "opaqueData": "xxx"
    },
    "subs": []
}
```
2. Then using `aws s3api put-bucket-notification-configuration --bucket mlp-deelooper --notification-configuration file://notification.xml  --endpoint-url http://10.9.8.102:80 --profile test3` to create bucket notification for a bucket.
```
[root@smb-test-1 deeproute]# cat notification.xml
{
    "TopicConfigurations": [
        {
            "Id": "test1",
            "TopicArn": "arn:aws:sns:os-dsglczutvqsgowpz::c-dev-prod-bucket-notification",
            "Events": [
                "s3:ObjectCreated:*",
                "s3:ObjectRemoved:*"
            ]
        }
    ]
}
```
3. After running for a few days, we found that `put_obj` operation all failed
![20230104-120152](https://user-images.githubusercontent.com/10202906/210482366-1bb53462-1332-4b85-bad2-8875cbe36945.jpg)
rgw log:
```
6448 debug 2022-12-27T07:31:48.329+0000 7f833becb700  1 beast: 0x7f83c48ba630: 10.3.8.33 - mlp [27/Dec/2022:07:31:48.304 +0000] "HEAD /mlp-data-warehouse HTTP/1.1" 200 0 - "aws-sdk-go/1.40.25 (go1.18.3; linux; amd64)" - latency=0.025000203s
6449 debug 2022-12-27T07:31:48.336+0000 7f832369a700  1 ====== starting new request req=0x7f83c48ba630 =====
6450 debug 2022-12-27T07:31:48.342+0000 7f82badc9700  1 req 750000557323249480 0.006000049s s3:put_obj ERROR: failed to reserve notification on queue: c-dev-prod-bucket-notification. error: -28
6451 debug 2022-12-27T07:31:48.343+0000 7f82badc9700  1 ====== req done req=0x7f83c48ba630 op status=-2218 http_status=503 latency=0.007000057s ======

```
[rgwlog.tar.gz](https://github.com/rook/rook/files/10341339/rgwlog.tar.gz)

ceph version: pacific 16.2.10

My colleague created an [ceph issue](https://tracker.ceph.com/issues/58375) to track it.

**Comment by yuvalif** [comment, marked as answer]
@microyahoo I don't think this is a rook issue. the problem is with the kafka broker and ceph (please see my reply in the ceph tracker).
as for the maturity of bucket notifications in rook, as @travisn said the feature is out there for a while, and we haven't observed any issues with it so far (note that rook only deals with the configuration of the topics and notifications).

**Reply by microyahoo** [reply]
Yes, it is not a rook issue, thanks for your nice clarification!

**Reply by travisn** [reply]
> @microyahoo I don't think this is a rook issue. the problem is with the kafka broker and ceph (please see my reply in the ceph tracker). as for the maturity of bucket notifications in rook, as @travisn said the feature is out there for a while, and we haven't observed any issues with it so far (note that rook only deals with the configuration of the topics and notifications).

Sounds great, let's consider moving the feature to stable with #11497.

**Reply by microyahoo** [reply]
Hi @yuvalif, how to handle or work around it when met this issue?

**Reply by yuvalif** [reply]
Hi, @microyahoo, what do you want to happen when your kafka broker is down?
if you don't mind that notifications are lost you can set `kafka-ack-level` to `none`.

**Reply by microyahoo** [reply]
We can tolerate notification loss, but not block `put_obj` operation.

**Reply by yuvalif** [reply]
so in this case, set `kafka-ack-level` to `none` at topic creation.
